### ISSUE TITLE
Trying to undo adding date from reconciled value gives error

### ISSUE THREAD

**Issue by ostephens** [OpenRefine/OpenRefine#1600]
When reporting a bug please provide the following information to help reproduce the bug:

#### Version of OpenRefine used (Google Refine 2.6, OpenRefine2.8, an other distribution?):
Switching between OpenRefine 2.8 and build from master branch or master branch + #1597 

#### Steps followed to create the issue:
In OpenRefine 2.8

- Reconcile a column to Wikidata
- Add a column using "Add columns from reconciled values" with a date type (e.g. 'date of birth').

In build from master branch or master branch + #1597 
Try to undo step

#### Current Results:
Fails with UI message:
Failed to load change file /Users/damyantiandowen/Library/Application Support/OpenRefine/2098864358506.project/history/1526070534696.change.zip

Log is:

16:18:02.950 [                   refine] POST /command/core/undo-redo (1808ms)
16:18:02.958 [                  command] Exception caught (8ms)
java.lang.RuntimeException: Failed to load change file /Users/damyantiandowen/Library/Application Support/OpenRefine/2098864358506.project/history/1526070534696.change.zip
	at com.google.refine.io.FileHistoryEntryManager.loadChange(FileHistoryEntryManager.java:82)
	at com.google.refine.history.HistoryEntry.revert(HistoryEntry.java:155)
	at com.google.refine.history.History.undo(History.java:236)
	at com.google.refine.history.History.undoRedo(History.java:176)
	at com.google.refine.history.HistoryProcess.performImmediate(HistoryProcess.java:82)
	at com.google.refine.process.ProcessManager.queueProcess(ProcessManager.java:97)
	at com.google.refine.commands.history.UndoRedoCommand.doPost(UndoRedoCommand.java:69)
	at com.google.refine.RefineServlet.service(RefineServlet.java:178)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:820)
	at org.mortbay.jetty.servlet.ServletHolder.handle(ServletHolder.java:511)
	at org.mortbay.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1166)
	at org.mortbay.servlet.UserAgentFilter.doFilter(UserAgentFilter.java:81)
	at org.mortbay.servlet.GzipFilter.doFilter(GzipFilter.java:132)
	at org.mortbay.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1157)
	at org.mortbay.jetty.servlet.ServletHandler.handle(ServletHandler.java:388)
	at org.mortbay.jetty.security.SecurityHandler.handle(SecurityHandler.java:216)
	at org.mortbay.jetty.servlet.SessionHandler.handle(SessionHandler.java:182)
	at org.mortbay.jetty.handler.ContextHandler.handle(ContextHandler.java:765)
	at org.mortbay.jetty.webapp.WebAppContext.handle(WebAppContext.java:418)
	at org.mortbay.jetty.handler.HandlerWrapper.handle(HandlerWrapper.java:152)
	at org.mortbay.jetty.Server.handle(Server.java:326)
	at org.mortbay.jetty.HttpConnection.handleRequest(HttpConnection.java:542)
	at org.mortbay.jetty.HttpConnection$RequestHandler.content(HttpConnection.java:938)
	at org.mortbay.jetty.HttpParser.parseNext(HttpParser.java:755)
	at org.mortbay.jetty.HttpParser.parseAvailable(HttpParser.java:212)
	at org.mortbay.jetty.HttpConnection.handle(HttpConnection.java:404)
	at org.mortbay.jetty.bio.SocketConnector$Connection.run(SocketConnector.java:228)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
Caused by: java.lang.reflect.InvocationTargetException
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at com.google.refine.history.History.readOneChange(History.java:83)
	at com.google.refine.history.History.readOneChange(History.java:69)
	at com.google.refine.io.FileHistoryEntryManager.loadChange(FileHistoryEntryManager.java:96)
	at com.google.refine.io.FileHistoryEntryManager.loadChange(FileHistoryEntryManager.java:80)
	... 29 more
Caused by: com.fasterxml.jackson.core.JsonParseException: Unrecognized token 'Mon': was expecting ('true', 'false' or 'null')
 at [Source: (String)"Mon Mar 10 19:00:00 EST 1952"; line: 1, column: 4]
	at com.fasterxml.jackson.core.JsonParser._constructError(JsonParser.java:1798)
	at com.fasterxml.jackson.core.base.ParserMinimalBase._reportError(ParserMinimalBase.java:673)
	at com.fasterxml.jackson.core.json.ReaderBasedJsonParser._reportInvalidToken(ReaderBasedJsonParser.java:2835)
	at com.fasterxml.jackson.core.json.ReaderBasedJsonParser._handleOddValue(ReaderBasedJsonParser.java:1889)
	at com.fasterxml.jackson.core.json.ReaderBasedJsonParser.nextToken(ReaderBasedJsonParser.java:747)
	at com.google.refine.model.ReconCandidate.loadStreaming(ReconCandidate.java:105)
	at com.google.refine.model.changes.DataExtensionChange.load(DataExtensionChange.java:461)
	... 37 more

#### Expected Results:
Should undo the column addition

**Comment by jackyq2015** [comment]
I can reproduce the issue. But it does not have to open it from the master version of OR. To reproduce on V 2.8:
1. add a date column from recon
2. shutdown the OR
3. Reopen the OR
4. undo the column just added

**Comment by ostephens** [comment]
Checked and I see the same behaviour. I'm guessing that until you restart OR the value is cached?

**Comment by jackyq2015** [comment]
yes, it is cached. Let me try V2.7 and report.

**Comment by ostephens** [comment]
I don't think that feature was in 2.7

**Comment by ostephens** [comment]
My guess is that it is something to do with the history writer on that feature not writing dates to the history correctly - but I'm about to get on a plane for 9hrs so can't look in more detail now

**Comment by jackyq2015** [comment]
@ostephens Right, just checked this feature is not in 2.7

**Comment by wetneb** [comment]
So I have looked into this and this bug comes from the original Freebase implementation that I have reused: the serialization and deserialization codes are just inconsistent. I'm trying to think of a fix that will both fix serialization for new changes while fixing somehow deserialization for the changes that have already been written.

**Comment by thadguidry** [comment]
@wetneb Can you check the Project version and if 2.7 or earlier run original, otherwise run newer ?
Perhaps yet another way is when we load the history file itself...you can inspect the file date and if earlier than "2017-11-18" consider it original...and when we write back, ensure it becomes newer ?

```
2.8
com.google.refine.model.changes.CellChange
row=0
cell=0
old={"v":"Autonomous single-door controller with built-in Proximity card reader and 3x4 keyboard for programming."}
new={"v":"{\n    \"id\": 1,\n    \"name\": \"A green door\",\n    \"price\": 12.50,\n    \"tags\": [\"home\", \"green\"]\n}"}
/ec/
```

**Comment by jackyq2015** [comment]
fixed by PR #1617
